src/awkward/_broadcasting.py

```diff
@@ -144,54 +143,13 @@ def ensure_common_length(inputs, options: BroadcastOptions) -> ShapeItem:
     return length
 
 
-def all_same_offsets(backend: Backend, inputs: list) -> bool:
-    index_nplike = backend.index_nplike
-
-    offsets = None
-    for x in inputs:
-        if isinstance(x, ListOffsetArray):
-            if offsets is None:
-                offsets = x.offsets.raw(index_nplike)
-            elif not index_nplike.array_equal(offsets, x.offsets.raw(index_nplike)):
-                return False
-
-        elif isinstance(x, ListArray):
-            starts = x.starts.raw(index_nplike)
-            stops = x.stops.raw(index_nplike)
-
-            if not index_nplike.array_equal(starts[1:], stops[:-1]):
-                return False
-
-            elif offsets is None:
-                offsets = index_nplike.empty(starts.shape[0] + 1, dtype=starts.dtype)
-                if offsets.shape[0] == 1:
-                    offsets[0] = 0
-                else:
-                    offsets[:-1] = starts
-                    offsets[-1] = stops[-1]
-
-            elif not index_nplike.array_equal(offsets[:-1], starts) or (
-                stops.shape[0] != 0 and offsets[-1] != stops[-1]
-            ):
-                return False
-
-        elif isinstance(x, RegularArray):
-            if x.size == 0:
-                my_offsets = index_nplike.empty(0, dtype=np.int64)
-            else:
-                my_offsets = index_nplike.arange(
-                    0, x.content.length + 1, x.size, dtype=np.int64
-                )
-
-            if offsets is None:
-                offsets = my_offsets
-            elif not index_nplike.array_equal(offsets, my_offsets):
-                return False
-
-        elif isinstance(x, Content):
-            return False
-
-    return True
+def offset_source_rank(list_content: ak.contents.Content) -> int:
+    if isinstance(list_content, ak.contents.ListOffsetArray):
+        return 0
+    elif isinstance(list_content, ak.contents.ListArray):
+        return 1
+    else:
+        raise AssertionError
```

Comment:
This might benefit from a better name. The purpose of this function is to score list-types in terms of the preference to use their offsets. ListOffsetArray can avoid an allocation (at the cost of reading the zeroeth value) to build offsets, whilst `ListArray` cannot. This might be premature optimisation; at one point I also scored regular arrays here, but that turned out to be a mistake.